woocommerce/woocommerce
# POC: Add WooCommerce block template class to handle extensibility

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR demonstrates how we could use a generic block template class, `WooCommerceBlockTemplate`, to allow better extensibility and parsing of all admin templates.

The API is intentionally simple to allow usage across not just the product editor, but most admin template cases and can be used like so:

```
$block_template = new WooCommerceBlockTemplate();
$block_template->add_block(
	array(
		'id'                 => 'my-block-2',
		'blockName' => 'my-block',
		'order'           => 10,
	)
);
$block_template->add_block(
	array(
		'id'                 => 'my-block-3',
		'blockName' => 'my-block',
		'parent'         => 'my-block-2',
	)
);
$block_template->add_block(
	array(
		'id'                 => 'my-block-1',
		'blockName' => 'my-block',
		'order'           => 5,
	)
);
$block_template->remove_block( 'test-block-1' );
$block_template->get_parsed_template();
```

We may want to make this class abstract and force consumers to implement methods that will be required later to register this template:

```
class MyAdminTemplate extends WooCommerceBlockTemplate {
	public function get_slug() {
		return 'my-admin-template';
	}
}
```

Closes # .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make a `GET` request to `wp-json/wp/v2/templates?post_type=woocommerce_product_editor_template`
2. See the parsed templates



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
